src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -153,71 +238,114 @@
+                final List<MarkDuplicatesSparkRecord> emptyFragments = stratifiedByType.get(MarkDuplicatesSparkRecord.Type.EMPTY_FRAGMENT);
+                final List<MarkDuplicatesSparkRecord> fragments = stratifiedByType.get(MarkDuplicatesSparkRecord.Type.FRAGMENT);
+                final List<Pair> pairs = (stratifiedByType.containsKey(MarkDuplicatesSparkRecord.Type.PAIR))?
+                        stratifiedByType.get(MarkDuplicatesSparkRecord.Type.PAIR).stream().map(p -> (Pair)p).collect(Collectors.toList())
+                        : null;
+                final List<MarkDuplicatesSparkRecord> passthroughs = stratifiedByType.get(MarkDuplicatesSparkRecord.Type.PASSTHROUGH);
+
+                //empty MarkDuplicatesSparkRecord signify that a pair has a mate somewhere else
+                // If there are any non-fragment placeholders at this site, mark everything as duplicates, otherwise compute the best score
+                if (Utils.hasElements(fragments) && !Utils.hasElements(emptyFragments)) {
+                    final Tuple2<IndexPair<String>, Integer> bestFragment = handleFragments(fragments);
+                    nonDuplicates.add(bestFragment);
+                }
 
-            List<GATKRead> out = Lists.newArrayList();
+                if (Utils.hasElements(pairs)) {
+                    nonDuplicates.add(handlePairs(pairs, finder));
+                }
 
-            // As in Picard, unpaired ends left alone.
-            for (final PairedEnds pair : paired.get(false)) {
-                out.add(pair.first());
+                if (Utils.hasElements(passthroughs)) {
+                    nonDuplicates.addAll(handlePassthroughs(passthroughs));
+                }
             }
 
-            // Order by score using ReadCoordinateComparator for tie-breaking.
-            Comparator<PairedEnds> pairedEndsComparator =
-                    Comparator.<PairedEnds, Integer>comparing(pe -> pe.score(scoringStrategy)).reversed()
-                            .thenComparing((o1, o2) -> new ReadCoordinateComparator(header).compare(o1.first(), o2.first()));
-            final List <PairedEnds> scored = paired.get(true).stream().sorted(pairedEndsComparator).collect(Collectors.toList());
-
-            final PairedEnds best = Iterables.getFirst(scored, null);
-            if (best == null) {
-                return out.iterator();
-            }
+            return nonDuplicates.iterator();
+        });
+    }
 
-            // Mark everyone who's not best as a duplicate
-            for (final PairedEnds pair : Iterables.skip(scored, 1)) {
-                pair.first().setIsDuplicate(true);
-                pair.second().setIsDuplicate(true);
-            }
+    // Note, this uses bitshift operators in order to perform only a single groupBy operation for all the merged data
+    private static long getGroupKey(MarkDuplicatesSparkRecord record) {
+        return record.getClass()==Passthrough.class?-1:
```

Comment:
did you open an issue to resolve the readgroup issue?

Reply:
yes #4700